### PR TITLE
WebAssembly + WebUSB support

### LINKED ISSUE #501: WebUSB Wasm Support
WebUSB tanked.

Please get this library running in Wasm so we can get USB support in all browsers.

### PR DESCRIPTION
**Pull request by RReverser**
(This is a follow-up from https://github.com/libusb/libusb/issues/501#issuecomment-1017750127.)

I implemented WebAssembly (via Emscripten) + WebUSB support in libusb as a separate backend specifically so that it would be easier to upstream the changes. There are a few changes outside of the backend's folder, but only where they were unavoidable.

In the process of making this port I've made various improvements to Emscripten itself to make porting of this and other similar libraries easier in the future, especially in area of async integration, embedding JavaScript snippets and more.

Afterwards, I published an article on how this port was made at https://web.dev/porting-libusb-to-webusb/ with a follow-up in https://web.dev/porting-gphoto2-to-the-web/, which show how the libusb fork was used almost "out of the box" with libgphoto2 to build a web UI and control DSLR cameras from the web.

Given the complexity and variety of operations of gPhoto2, I'm hoping that such testing was extensive enough to eliminate most of the critical issues by now, but I wouldn't be surprised if there are more. As suggested by @mcuee in https://github.com/libusb/libusb/issues/501#issuecomment-1018403790, I'm opening this PR to get more eyes on it and have more people test it for now, with the hope that it can be merged upstream afterwards.

I have instructions for port usage in https://web.dev/porting-libusb-to-webusb/#using-the-port, but the TL;DR is:
 - point application to use the local checkout of the libusb fork
 - configure the libusb fork with `emconfigure`/`emmake` wrappers from [Emscripten toolchain](https://github.com/emscripten-core/emsdk) and `--host=wasm32`
 - build & link the application with `--bind -s ASYNCIFY -s ALLOW_MEMORY_GROWTH -pthread`
 - if you're building without optimizations, sometimes you'll get weird errors from WebAssembly due to insufficient Asyncify stack space; to increase it, add `-s ASYNCIFY_STACK_SIZE=10485760` (for example this would set it 10MB which should be more than enough for most apps, and it's only for debug builds anyway).

Let me know if there are any questions or concerns.

Fixes #501.

### CODE REVIEW

**Comment by morsssss** [comment]
I've been following this project with great interest... it could make it much easier to create libusb-based drivers for the Web!

Does someone have time to take a look at this PR?

**Comment by orangecms** [comment]
> I've been following this project with great interest... it could make it much easier to create libusb-based drivers for the Web!
> 
> Does someone have time to take a look at this PR?

Hi, I'm not a contributor to libusb (yet, heh), but I'm building on top of this work in order to bring https://github.com/sigrokproject/libsigrok to the web and reuse it for a web version of PSLab Desktop; see https://github.com/fossasia/pslab-desktop/issues/697.

I will add some more feedback here.

**Comment by mcuee** [comment]
Good comments from Tormod here. Hopefully more people will jump in to review and provide feedbacks or report testing results.

> OTOH #1057 is a non-invasive patch so easier to merge, but I'd like to see more testing and reviews from people using it, to make sure it is mature before landing in master.

**Comment by BatslyAdams** [comment]
Hi! I also wanted to report success here with the fork using the USB peripheral on a custom built ATMEGA164a board. The device was originally controlled via a CLI tool written in C, after retargeting to Emscripten using the [latest commit](https://github.com/RReverser/libusb/tree/2db3897d7e642a0993dd881c4d9ff60dfc9bf53a) the device correctly enumerates in Windows 10 Chrome and has demonstrated 100% functionality against the original binary when calling main with args from JS. 

I'd be happy to provide any further technical details, but I wanted to say that this feature has directly solved the long-standing issue of being able to provide an up to date cross-platform toolchain for users of this device.

**Comment by hjelmn** [comment]
This needs to be squashed down to a much smaller number of commits before it can be properly reviewed.

**Comment by RReverser** [comment]
I don't think you need to review commit-by-commit - as noted above, most of the changes are self-contained and could as well be squashed into a single commit, so it's one of those cases where it's easier to just review the patch as a whole.

_(Side-note: while I'm starting to ramp up and doing some minor things again, as a Ukrainian, I'm still a lot less active than usual and will be somewhat slow to respond to feedback. Please feel free to leave feedback regardless, just saying I haven't been looking at code lately and only starting to get in the right mindset again.)_

**Comment by RReverser** [comment]
Ok, so do you want me to just squash everything before reviewing or..? As more people start using this for their own projects, I'd love to see the backend merged into libusb for easier discoverability, rather than having to keep my own fork up-to-date.

**Comment by qodeos** [comment]
Do not use git to do that. Re-open this commit, and create it using GitHub.

**Comment by RReverser** [comment]
That was a question to @hjelmn.

**Comment by mcuee** [comment]
@RReverser
Please help to fix the github warning.
> Merging is blocked
> The base branch requires all commits to be signed.

File: libusb/os/events_posix.c [hunk 1]
```diff
@@ -28,6 +28,34 @@
 #ifdef HAVE_TIMERFD
 #include <sys/timerfd.h>
 #endif
+#ifdef __EMSCRIPTEN__
+// On Emscripten pipe does not conform to the spec and does not block until events are available,
+// which makes it unusable for event system and often results in deadlocks when `pipe` is in a loop
+// like it is in libusb.
+//
+// To work around that, I'm introducing a custom event system based on browser event emitters.
```
Comment (tormodvolden) [inline libusb/os/events_posix.c:36]: Comment should use /* */ format.

**Comment by tormodvolden** [comment]
About squashing: If there are parts that can potentially be useful without the rest, they should be in separate commits. Examples could be a rework of existing code that improves this code independently, or a new feature that is useful also without a second feature. From what I see in this PR I guess squashing it to one commit is best since everything is depending on all other parts. The long list of fix-up commits and work progress is not useful for reviewing, only if there are fix-ups after review comments that are better evaluated on their own. If this is still a WIP and people are testing it at certain revisions, it would IMO still make sense to preserve the development history in separate commits until it is considered ready for merging. And true, e.g. GitHub offers the "Files changed" view so we don't need to look at the commits to see the code changes. But we need to review the final commit message in the end. Files like AUTHORS we generally update as part of the release process so it is not needed here but this is only a question of consistency and nothing I have strong opinion about.

**Comment by RReverser** [comment]
> About squashing: If there are parts that can potentially be useful without the rest, they should be in separate commits. Examples could be a rework of existing code that improves this code independently, or a new feature that is useful also without a second feature. From what I see in this PR I guess squashing it to one commit is best since everything is depending on all other parts. The long list of fix-up commits and work progress is not useful for reviewing, only if there are fix-ups after review comments that are better evaluated on their own.

I agree; I did some interim squashing as I was working on this fork as I also usually prefer to keep the history clean, but in the very end for this particular PR it makes more sense to squash everything into one commit.

My only suggestion is that I do that after addressing any feedback on this PR, everything in one go.

**Comment by RReverser** [comment]
> And true, e.g. GitHub offers the "Files changed" view so we don't need to look at the commits to see the code changes.

Btw, aside from Github UI - for people who prefer to review patch files, you can also get everything together by appending `.patch` to URL, e.g. for this PR it would be https://github.com/libusb/libusb/pull/1057.patch

File: libusb/os/events_posix.c (hunk 1 above)
Comment (RReverser) [reply libusb/os/events_posix.c]: Done

**Comment by RReverser** [comment]
> Is this a limitation of the approach or am I misunderstanding things?

It's definitely a limitation, but rather of the web / JS model itself. On the web you can't just access JS objects (which WebUSB is also represented with) from multiple threads. You can serialize objects and post to other threads and deserialize back on that side, but that's not the same and doesn't fit well into C++ threading model.

Your best bet would be doing a (hopefully, small) refactor to make driver do all the USB work from the same thread.

> Also, I guess some of the flags used in the PR might not be necessary for the library build (` -s INVOKE_RUN=0 -s EXPORTED_RUNTIME_METHODS=['callMain']`), no?

Yeah those flags don't affect the library build since it's just a static archive and not linked.

**Comment by RReverser** [comment]
@tormodvolden Anything else in terms of feedback / what's next for this PR?

**Comment by bastibl** [comment]
Before testing this PR, I used a [webusb-libusb shim](https://github.com/luigifcruz/webusb-libusb) for said driver, which worked. So it should be possible. My understanding of [this](https://github.com/emscripten-core/emscripten/pull/13315) discussion is that it should be possible with [this](https://github.com/emscripten-core/emscripten/pull/15737) extension.

**Comment by RReverser** [comment]
> My understanding of [this](https://github.com/emscripten-core/emscripten/pull/13315) discussion is that it should be possible with [this](https://github.com/emscripten-core/emscripten/pull/15737) extension.

IIRC, that API basically proxies work back from any thread to the main thread. It helps with more seamless porting, but kind of defeats the main benefit of separate threads as the main thread is still doing all the work, which might lead to unexpected behaviour / performance results. I'm not sure how I feel about it...

**Comment by reillyeon** [comment]
The design of libusb assumes that the platform API is safe to call from multiple threads as long as a lock is held. JavaScript APIs on the other hand are "thread hostile" and really need to be used consistently from the same thread.

I think the only solution is to either design the WebUSB backend so that it binds to a particular thread (probably the one where it was initialized) but works from other threads by posting messages, or by having it create it's own thread where all the JavaScript API calls are made. The advantage of the former is that it'll avoid overhead when used from a single thread but could deadlock if that thread is busy.

Fixing this from the WebUSB side would require a way to "clone" a WebUSB device object and pass it to another worker. I've been thinking about how transferring devices between workers should work and this isn't an option I was considering but given this use case I'll see if it makes sense.

**Comment by RReverser** [comment]
@reillyeon Thanks for chiming in!

From Emscripten perspective, I also have some ideas on how sharing `val` with other threads could be made possible more seamlessly in the future, but, given the complexity on all sides, I'm definitely keen on leaving that outside of scope of this PR.

**Comment by RReverser** [comment]
@tormodvolden Sorry, but... another ping. I'd really like to get this unblocked, it's been some time :)

**Comment by tormodvolden** [comment]
I appreciate that users of this code has commented on it and that there has been a discussion on the design. It seems like there are no immediate plans to change anyway though. So I am in favour of merging this now, and see if more people will use it and further develop and maintain it.

In view of comment https://github.com/libusb/libusb/pull/1057#issuecomment-1139470444 do you need the `AC_SUBST(AM_LDFLAGS)` ? The CI builds fail on Linux, I don't know if it is related to this, but it seems like the only change in common code.

**Comment by RReverser** [comment]
> In view of comment [#1057 (comment)](https://github.com/libusb/libusb/pull/1057#issuecomment-1139470444) do you need the `AC_SUBST(AM_LDFLAGS)` ?

I do need `AC_SUBST` to correctly link executables for examples & tests, and, as mentioned above, `AM_LDFLAGS` doesn't affect the static library anyway. I can scope it only to Emscripten build though if you think that would help.

> The CI builds fail on Linux, I don't know if it is related to this, but it seems like the only change in common code.

I doubt it's related because on the initial commit the Linux build did pass:
image

It only started failing when I updated comment styles, which is unlikely to be the culprit, and I suspect it's rather something upstream that has changed:
image

I was hoping you'd know what that change could be, but, if not, I can try to dig in myself.

**Comment by RReverser** [comment]
Let's see if merging latest master helps...

**Comment by RReverser** [comment]
> Let's see if merging latest master helps...

It doesn't... I'll play with this a bit.

**Comment by RReverser** [comment]
So I just tried building it locally, and the build on Linux still succeeds. On the CI it's `umockdev test` task that's failing with

> /usr/bin/ld: cannot find ../libusb/.libs/libusb-1.0.so: No such file or directory"

but it's built successfully, too. I'm not sure what's going on on the CI :/

**Comment by RReverser** [comment]
Comparing CI logs, one difference I see is that on master libusb-1.0.la is being linked with `CCLD`, whereas in my fork for some reason it chooses `CXXLD`:

```
  CXXLD    libusb-1.0.la
```

I don't understand either why would that change (there's nothing suspicious in my changes to `configure.ac` I think, so maybe it chooses based on merely presence of a new C++ file? but that's unused for Linux target?), or whether that's the culprit. So weird :/

**Comment by RReverser** [comment]
> maybe it chooses based on merely presence of a new C++ file

Oh. Apparently this is it and kinda known automake problem: https://www.gnu.org/software/automake/manual/html_node/How-the-Linker-is-Chosen.html (example: https://github.com/mono/mono/issues/12435). TIL.

**Comment by RReverser** [comment]
Ok that was it. I copied the linker selection workaround from Haiku backend in Makefile.am (since, like Emscripten, it also uses C++) and Linux build on CI now passes again.

**Comment by tormodvolden** [comment]
Thanks for figuring that out!

**Comment by RReverser** [comment]
Thanks for merging! Initially, when planning to do the final squash, I thought I'd keep either the instructions or at least the URL from this part:

> I have instructions for port usage in [web.dev/porting-libusb-to-webusb/#using-the-port](https://web.dev/porting-libusb-to-webusb/#using-the-port), but the TL;DR is:
> 
> * point application to use the local checkout of the libusb fork
> * configure the libusb fork with `emconfigure`/`emmake` wrappers from [Emscripten toolchain](https://github.com/emscripten-core/emsdk) and `--host=wasm32`
> * build & link the application with `--bind -s ASYNCIFY -s ALLOW_MEMORY_GROWTH -pthread`
> * if you're building without optimizations, sometimes you'll get weird errors from WebAssembly due to insufficient Asyncify stack space; to increase it, add `-s ASYNCIFY_STACK_SIZE=10485760` (for example this would set it 10MB which should be more than enough for most apps, and it's only for debug builds anyway).

If not, do you think there's some other place where we should add at least the URL to the blog post, so that other users can find the instructions from libusb repo more easily?

**Comment by RReverser** [comment]
I guess I can add something to https://github.com/libusb/libusb/wiki#supported-environments.

**Comment by RReverser** [comment]
UPD: done.

image

**Comment by morsssss** [comment]
Sounds like a nice case for a followup PR...

**Comment by RReverser** [comment]
I added support for multithreading via proxying + lots of other fixes in https://github.com/libusb/libusb/pull/1339. Please give it a try!

**Pull request closed without merging**